File: src/parsing/operations/create_node.ts
import Database from "../../graph/database";
import Node from "../../graph/node";
import ASTNode from "../ast_node";
import Operation from "./operation";

class CreateNode extends Operation {
    private _node: Node | null = null;
    private _statement: ASTNode | null = null;
    private _isStatic: boolean = false;
    private _refreshEveryMs: number | null = null;
    constructor(
        node: Node,
        statement: ASTNode,
        isStatic: boolean = false,
        refreshEveryMs: number | null = null
    ) {
        super();
        this._node = node;
        this._statement = statement;
        this._isStatic = isStatic;
        this._refreshEveryMs = refreshEveryMs;
    }
    public get node(): Node | null {
        return this._node;
    }
    public get statement(): ASTNode | null {
        return this._statement;
    }
    public get isStatic(): boolean {
        return this._isStatic;
    }
    public get refreshEveryMs(): number | null {
        return this._refreshEveryMs;
    }
    public run(): Promise<void> {
        return new Promise(async (resolve, reject) => {
            try {
                if (this._node === null) {
                    throw new Error("Node is null");
                }
                const db: Database = Database.getInstance();
                db.addNode(this._node, this._statement!, this._isStatic, this._refreshEveryMs);
                resolve();
            } catch (error) {
                reject(error);
            }
        });
    }
    public get results(): Record<string, any>[] {
        return [];
    }
}

export default CreateNode;

File: flowquery-py/src/parsing/operations/create_node.py
"""Represents a CREATE operation for creating virtual nodes."""

from typing import Any, Dict, List, Optional

from ...graph.database import Database
from ...graph.node import Node
from ..ast_node import ASTNode
from .operation import Operation


class CreateNode(Operation):
    """Represents a CREATE operation for creating virtual nodes."""

    def __init__(
        self,
        node: Node,
        statement: ASTNode,
        is_static: bool = False,
        refresh_every_ms: Optional[int] = None,
    ) -> None:
        super().__init__()
        self._node = node
        self._statement = statement
        self._is_static = is_static
        self._refresh_every_ms = refresh_every_ms

    @property
    def node(self) -> Node:
        return self._node

    @property
    def statement(self) -> ASTNode:
        return self._statement

    @property
    def is_static(self) -> bool:
        return self._is_static

    @property
    def refresh_every_ms(self) -> Optional[int]:
        return self._refresh_every_ms

    async def run(self) -> None:
        if self._node is None:
            raise ValueError("Node is null")
        db = Database.get_instance()
        db.add_node(
            self._node,
            self._statement,
            self._is_static,
            self._refresh_every_ms,
        )

    @property
    def results(self) -> List[Dict[str, Any]]:
        return []
